File: 3.Longest Substring Without Repeating Characters/main.py




class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        set_temp = set()
        i,j = 0,0
        max = 0
        n = len(s)
        while i < n and j < n:
            if s[i] not in set_temp:
                set_temp.add(s[i])
                i = i + 1
                if len(set_temp) > max:
                    max = len(set_temp)

            else:
                while s[i] in set_temp:
                    set_temp.remove(s[j])
                    j = j + 1
        return max






if __name__ == "__main__" :
    s = Solution()
    string = "pwwkew"
    print(s.lengthOfLongestSubstring(string))

File: 3.Longest Substring Without Repeating Characters/main.cpp
#include<cstring>
#include<iostream>
#include<set>
#include<algorithm>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // 哈希集合，记录每个字符是否出现过
        set<char>set_temp;
        int i = 0, j = 0, result = 0;
        int n = s.length();
        while (i < n && j < n){
            if(set_temp.count(s[i]) == 0){
                set_temp.insert(s[i]);
                i += 1;
                result = max((int)set_temp.size(), result);
                if (set_temp.size() > result){
                    result = set_temp.size();
                }
            }
            else{
                while(set_temp.count(s[i])){
                    set_temp.erase(s[j]);
                    j += 1;
                }
            }
        }
        return result;
    }
};

int main(){
    Solution s;
    string ss = "pwwkew";
    // cout << s.lengthOfLongestSubstring(ss) << endl;
    cout << s.lengthOfLongestSubstring(ss) << endl;
    return 0;
}